core/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java

```diff
@@ -82,48 +99,152 @@ public void startedPrimaryReinitialized(ShardRouting startedPrimaryShard, ShardR
+            int upperBound = oldIndexMetaData.getNumberOfReplicas() + 1; // +1 for the primary
+            int overflow = inSyncAllocationIds.size() - upperBound;
+            if (overflow > 0) {
+                // trim entries that have no corresponding shard routing in the cluster state (i.e. trim unavailable copies)
+                Set<String> assignedAllocations = newRoutingTable.shardRoutingTable(shardId).assignedShards().stream()
+                    .map(shr -> shr.allocationId().getId()).collect(Collectors.toSet());
+                Set<String> candidatesToRemove = Sets.difference(inSyncAllocationIds, assignedAllocations);
+                for (Iterator<String> it = candidatesToRemove.iterator(); it.hasNext() && overflow > 0; ) {
+                    inSyncAllocationIds.remove(it.next());
+                    overflow--;
+                }
+                assert overflow == 0 :
+                    "not enough candidates to remove, candidates: " + candidatesToRemove + " overflow: " + overflow;
+            }
+
+            // only update in-sync allocation ids if there is at least one entry remaining. Assume for example that there only
+            // ever was a primary active and now it failed. If we were to remove the allocation id from the in-sync set, this would
+            // create an empty primary on the next allocation (see ShardRouting#allocatedPostIndexCreate)
+            if (inSyncAllocationIds.isEmpty() && oldInSyncAllocationIds.isEmpty() == false) {
+                assert updates.firstFailedPrimary != null :
+                    "in-sync set became empty but active primary wasn't failed: " + oldInSyncAllocationIds;
+                if (updates.firstFailedPrimary != null) {
+                    // add back allocation id of failed primary
+                    inSyncAllocationIds.add(updates.firstFailedPrimary.allocationId().getId());
+                }
+            }
+
+            // be extra safe here and only update in-sync set if it is non-empty
+            if (inSyncAllocationIds.isEmpty() == false) {
+                assert inSyncAllocationIds.isEmpty() == false;
+                if (indexMetaDataBuilder == null) {
+                    indexMetaDataBuilder = IndexMetaData.builder(oldIndexMetaData);
+                }
+                indexMetaDataBuilder.putInSyncAllocationIds(shardId.id(), inSyncAllocationIds);
+            }
+        }
+        return indexMetaDataBuilder;
+    }
+
+    /**
+     * Removes allocation ids from the in-sync set for shard copies for which there is no routing entries in the routing table.
+     * This method is called in AllocationService before any changes to the routing table are made.
+     */
+    public static ClusterState removeStaleIdsWithoutRoutings(ClusterState clusterState, List<StaleShard> staleShards) {
+        MetaData oldMetaData = clusterState.metaData();
+        MetaData.Builder metaDataBuilder = null;
+        // group staleShards entries by index
```

Comment:
can we assert that none of the `staleShards` entries exist as active entries in the routing table?